README.MD
just an example repository to test the AndroidReleaseBuildScript.sh
blatest blatest


ListJSON/changeLog.txt
### Change log of version  generated at Sat May 23 19:05:57 CEST 2015 ###
-----------------------------------------------------------------
	





ListJSONtest/src/me/blobb/listjson/test/FileInfo.java
package me.blobb.listjson.test;

import java.util.Date;

/**
 * @author andre.boddenberg@gmx.de
 * 
 * The FileInfo class provides <i>one</i> constructor to instantiate a FileInfo object.<br>
 * A FileInfo object can be used to handle the information of folders and files.<br><br>
 * Check the parameters for more details.<br><br>
 * 
 * Besides the typical getter and setter this class provides methods, which transform the long values<br>
 * of size and date into representative String (i. e. size=>"12.5 kB"  date=>"11 Sep 1987")
 * 
 * @param size		long -> size of the FileInfo object
 * @param date		long (1.1.1970) -> date of the FileInfo object.
 * @param name		String -> name of the FileInfo object.
 * @param folder	boolean -> variable to indicate whether the FileInfo object is a folder or a file.
 */
public class FileInfo
{
	private long size;
	private long date;
	private String name;
	private boolean folder;
	
	/**
	 * Constructor to instantiate a FileInfo object.<br>
	 * There is no default Constructor, so you <i>must</i> instantiate <br>
	 * a FileInfo object with <i>all four</i> parameters.
	 * 
	 * @param size		long -> size the FileInfo object should have.
	 * @param date		long (1.1.1970) -> date the FileInfo object should have.
	 * @param name		String -> name the FileInfo object should have.
	 * @param folder	boolean -> set true if FileInfo should be a folder.
	 */
	public FileInfo(long size, long date, String name,  boolean folder)
	{
		this.name = name;
		this.size = size;
		this.date = date;
		this.folder = folder;
	}

	/**
	 * method to get a representative String of the FileInfo object.<br><br>
	 * NOTE: if you want to implement a toJSON method you could take this as a template.
	 * 
	 * @return returns a representative String of the FileInfo (may useful for testing).
	 */
	public String toString()
	{
		return "name: " + this.name + 
				", date(long): " + String.valueOf(this.date) + 
				", size (in Byte): " + this.size + 
				", is folder = " + this.isFolder();
	}
	
	/**
	 * getter: gets the size of the FileInfo object as long
	 * 
	 * @return returns the size of the FileInfo object as long
	 */
	public long getRawSize()
	{
		return size;
	}
	
	/**
	 * Method to get a representative String of the size of the FileInfo object.<br>
	 * This method divides the FileInfo size(long) by 1024 until it's smaller than 1024.<br>
	 * While it counts each division and use it as an indicator for the "unit-prefix".<br>
	 * Afterwards it rounds the size to one digit accuracy.<br><br>
	 * 
	 * NOTE: if you consider implementing "Locale" to get date in correct language, 
	 * you could use it to define whether to show a dot or a comma.
	 * 
	 * @return returns a representative String of the size of the FileInfo object (i. e. "12.5 kB")
	 */
	public String getAcurateSize()
	{
		double dSize = (double) this.size;			// casting size to double to have a floating point variable
		String dimStr = null;						// String for the "unit-prefix"
		int dim = 0;								// counts the times of dividing

		// divide size until it's smaller than 1024
		while(dSize > 1024.0D)					
		{
			dSize = dSize/1024.0D;
			dim++;
		}
		
		// evaluate the "division-counter"
		switch (dim)							
		{
			case 0:	dimStr =" B";
					break;
			case 1:	dimStr =" kB";
					break;
			case 2:	dimStr =" MB";
					break;
			case 3: dimStr =" GB";
					break;
			case 4: dimStr =" TB";
					break;
			case 5:	dimStr =" PB";
					break;
			case 6: dimStr =" EB";
		}
		
		// round to one digit accuracy
		dSize = (double)Math.round(dSize * 10) / 10;		
		
		return String.valueOf(dSize) + dimStr;
	}

	/**
	 * set the size(long) of the FileInfo object
	 * 
	 * @param size pass size of FileInfo object as long
	 */
	protected void setRawSize(long size)  
	{
		this.size = size;
	}
	
	/**
	 * getter: date of FileInfo object as long
	 * 
	 * @return returns the date as long (1.1.1970)
	 */
	public long getRawDate()
	{
		return date;
	}

	/**
	 * This method turns the date variable(long) of the FileInfo object into a representative String<br><br>
	 * 
	 * NOTE: This method can be improved a lot. Especially if you want to implement a possibility to 
	 * 		 choose what "kind/types" of the date should be displayed (and in which language).
	 * 
	 * @return returns the date of the FileInfo object as a String "DD MMM YYYY".
	 */
	public String getAcurateDate(){
		Date df = new Date(this.date * 1000);
		
		String dStr = df.toString();
		
		//String dayWord = dStr.substring(0, dStr.indexOf(" ")); 	    not used in the ListJSONTest 
		dStr = dStr.substring(dStr.indexOf(" ")+1);
		
		String monthWord = dStr.substring(0, dStr.indexOf(" "));
		dStr = dStr.substring(dStr.indexOf(" ")+1);
		
		String dayNumber = dStr.substring(0, dStr.indexOf(" "));
		dStr = dStr.substring(dStr.indexOf(" ")+1);
		
		//String time =  dStr.substring(0, dStr.indexOf(" "));			not used in the ListJSONTest 
		dStr = dStr.substring(dStr.indexOf(" ")+1);
				
		//String timezone = dStr.substring(0, dStr.indexOf(" "));		not used in the ListJSONTest 
		dStr = dStr.substring(dStr.indexOf(" ")+1);
		
		String year = dStr;
		
		return dayNumber + " " + monthWord + " " + year;
	}
	
	/**
	 * setter: sets the date of the FileInfo object
	 * 
	 * @param date pass new date as long (1.1.1970)
	 */
	protected void setRawDate(long date)
	{
		this.date = date;
	}
	
	/**
	 * getter: returns the name of the FileInfo object as a String
	 * 
	 * @return name of the FileInfo object
	 */
	public String getName()
	{
		return name;
	}
	
	/**
	 * setter: sets the name of the FileInfo object
	 * 
	 * @param name pass the name of FileInfo object (folder/file) as String
	 */
	protected void setName(String name)
	{
		this.name = name;
	}

	/**
	 * method to get information about file or folder of FileInfo object.
	 * 
	 * @return folder returns true if FileInfo object is a folder 
	 */
	public boolean isFolder()
	{
		return folder;
	}

	/**
	 * setter: set the the value of folder to "true" or "false".
	 * 
	 * @param folder
	 */
	protected void setFolder(boolean folder)
	{
		this.folder = folder;
	}

}


ListJSONtest/src/me/blobb/listjson/test/MainActivityTest.java
package me.blobb.listjson.test;

import android.test.ActivityTestCase;
//import ListJSON.*;

/**
 * This is a simple framework for a test of an Application.  See
 * {@link android.test.ApplicationTestCase ApplicationTestCase} for more information on
 * how to write and extend Application tests.
 * <p/>
 * To run this test, you can type:
 * adb shell am instrument -w \
 * -e class me.blobb.listjson.MainActivityTest \
 * com.cortado.listjsontest.tests/android.test.InstrumentationTestRunner
 */
public class MainActivityTest extends ActivityTestCase {

    public MainActivityTest() {
    }
    
    public void testSomething(){
    	FileInfo finfo = new FileInfo(4954967l,49375935l,"foobar", false);
    	assertEquals(false, finfo.isFolder());
     	assertEquals("foobar", finfo.getName());	
    }
    
    public void testSomethingElse(){
    	FileInfo finfo = new FileInfo(4954967l,49375935l,"foobar", false);
    	assertEquals(false, finfo.isFolder());
     	assertEquals("foo", finfo.getName());
    }
}


ListJSON/debugBuild.sh
#!/bin/bash
#
# This is an easy debug build script,
# so all developers enable the debuggable mode
# in the AndroidManifest-xml, build the debug apk 
# and sign the debug apk with the same keystore,
# which is stored inside the repository.
#
# (PLEASE DO NOT PROVIDE THE RELEASE KEYSTORE HERE)
#
# @author: andre@blobb.me
# @date: 20.12.2015
#
# Please define the following parameters...
#
APPNAME="MainActivity"
KEYSTORE="ListJSONdebug.keystore"
ALIAS="blobb"
STOREPASS="password"
MANIFEST_PATH="AndroidManifest.xml"
################################################################################
set -e

setDebuggableToTrue() {
	sed -i 's/debuggable *= *"false"/debuggable="true"/;' $MANIFEST_PATH
	
	# checking changes
	check="true"
	
	debuggable_test=`grep debuggable $MANIFEST_PATH | sed 's/.*debuggable="//;s/".*//'`
	
	# checking debuggable flag 
	if [ "$debuggable_test" != "true" ]; then
			echo "ERROR: changing the AndroidManifest.xml"
			exit 1
	fi

	echo "1) changed debuggable attribute to \"true\" in ${MANIFEST_PATH}"

}

# check if manpage was asked
if  [ $# == 1 ] && [ "$1" == "-help" ]; then

	echo -e "\nThis is an easy debug build script,
so all developers enable the debuggable mode
in the AndroidManifest.xml, build the debug apk 
and sign the debug apk with the same keystore,
which is stored inside the repository.\n
Just invoke this script without any parameter,
or pass one random parameter for verbose mode.\n"

	exit
fi

buildLang=`ls | grep build.xml`

if [ $# == 0 ]; then
	echo -e "\n########## building Android debug version ##########\n"
	setDebuggableToTrue 

	if [ ${#buildLang} -gt 0 ]; then
		echo "2) building Android apk with ANT"
		invis=`ant debug` 
	else
		echo "2) building Android apk with GRADLE"
		invis=`./gradlew assembleDebug`
	fi

	echo -e "3) signing apk with ${KEYSTORE}\n"
	invis=`jarsigner -verbose -sigalg SHA1withRSA -digestalg SHA1 -keystore ${KEYSTORE} -storepass ${STOREPASS} bin/${APPNAME}-debug.apk ${ALIAS}`

	read  -p "Escape (CTRL+C) or press enter to install debug apk on device/emulator..."
	adb install -r bin/${APPNAME}-debug.apk

else
	echo -e "\n########## building Android debug version ##########"
	setDebuggableToTrue

	if [ ${#buildLang} -gt 0 ]; then
		echo -e "\n########## using ANT to build debug apk ##########\n"
		ant debug 
	else
		echo -e "\n########## using GRADLE to build debug apk #########\n"
		./gradlew assembleDebug 
	fi

	echo -e "\n########## signing apk with debugStore ##########\n"
	jarsigner -verbose -sigalg SHA1withRSA -digestalg SHA1 -keystore ${KEYSTORE} -storepass ${STOREPASS} bin/${APPNAME}-debug.apk ${ALIAS} 
	echo -e "\n########## debug build done ##########\n"

	if [ "$1" != "jenkins" ]; then
		read  -p "Escape (CTRL+C) or press enter to install debug apk on device/emulator."
		adb install -r bin/${APPNAME}-debug.apk
	fi
fi


ListJSON/src/me/blobb/listjson/MainActivity.java
package me.blobb.listjson;

import java.io.IOException;
import java.io.InputStream;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import me.blobb.listjson.R;

import android.app.Activity;
import android.os.Bundle;
import android.widget.ListView;

/**
 * @author andre.boddenberg@gmx.de
 * 
 * The MainActivity class provides a simple Activity with one ListView.<br>
 * The class instantiate a listView, parses a JSONfile to a representative String<br>
 * and parses this String to a FileInfo[]. <br><br>
 * 
 * The objects of the FileInfo[] are each displayed via an ArrayAdapter in a rowView (FileInfoArrayAdapter).<br>
 * For more details read the FileInfoArrayAdapter doc. 
 * 
 */
public class MainActivity extends Activity 
{
	@Override
	protected void onCreate(Bundle savedInstanceState) 
	{
		// call super Constructor and set View layout
		super.onCreate(savedInstanceState);							
	    setContentView(R.layout.list_view);
	    
	    // instantiate the default android list view
	    ListView lv = (ListView) findViewById(R.id.listview);
		
	    // getting file.json and creating FileInfo[]
		String json = parseJSONStringFromAssets("files.json");							
		FileInfo[] arFi = parseJSONStringToFileInfoArray(json);
		
		// instantiate a new FileInfoArrayAdapter to display FileInfo[] objects via list_view
	    FileInfoArrayAdapter adapter = new FileInfoArrayAdapter(this, R.layout.list_view, arFi);
	    lv.setAdapter(adapter);
	  }
	
	/**
	 * Method parses a JSONFile to a String and returns it.
	 * 
	 * @param file Name of a JSONfile that should be parsed to a String.<br>
	 * NOTE: If file is in a subfolder of /assets/.. you must pass the path.
	 * 
	 * @return Returns representative String of the parsed JSONfile.
	 */
	private String parseJSONStringFromAssets(String file) 
	{	
		InputStream is;
		byte[] buffer = null;
		
		try 
		{
			is = getAssets().open(file);
			
			buffer = new byte[is.available()];
			
			is.read(buffer);
			is.close();
			
			return new String(buffer, "UTF-8");
		
		} catch (IOException e) {
			e.printStackTrace();
			return null;
		}
	}

	/**
	 * Method to parse the JSONString to a FileInfo Array
	 * 
	 * @param  jsonString representative String of a JSONfile
	 * @return Returns a FileInfo[], which contains all objects of the folder and file Array in the passed jsonString.
	 */
	private FileInfo[] parseJSONStringToFileInfoArray(String jsonString)
	{
		FileInfo[] arFi = null;	
		
		try 
		{
			//instantiating JSONObject and jump into the root object
			JSONObject jsParent = new JSONObject(jsonString);
			jsParent = jsParent.getJSONObject("folder");				
			
			// creating JSONArray of the arrays folder and file
			JSONArray jsArFolders = jsParent.getJSONArray("folder");	// creating array
			JSONArray jsArFiles =  jsParent.getJSONArray("file");		// creating array

			// re-instantiating arFi and setting its length
			arFi = new FileInfo[jsArFolders.length() + jsArFiles.length()];
			
			// temporary JSONObject to work through every JSONObject inside the two arrays.
			JSONObject temp = null;
			
			// creating FileInfo objects of each JSONObjects inside "JSON folder array" 
			// and add it to arFi. 
			for(int i=0; i<jsArFolders.length(); i++)
			{
				temp = jsArFolders.getJSONObject(i);
				arFi[i] = new FileInfo(temp.getLong("size"), temp.getLong("date"), temp.getString("name"), true); 
			}
			
			// creating FileInfo objects of each JSONObjects inside "JSON file array" 
			// and add it to arFi.
			for(int i=0; i<jsArFiles.length() ; i++)
			{
				temp = jsArFiles.getJSONObject(i);
				arFi[i + jsArFolders.length()] = new FileInfo(temp.getLong("size"), temp.getLong("date"), temp.getString("name"), false); 
			}
		} catch (JSONException e) {
			e.printStackTrace();
		}
	  	return arFi;  	
	}
}
	

ListJSON/src/me/blobb/listjson/FileInfoArrayAdapter.java
package me.blobb.listjson;

import me.blobb.listjson.R;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

/**
 * @author andre.boddenberg@gmx.de
 * 
 * The class FileInfoArrayAdapter is an ArrayAdapter class to handle FileInfo Arrays.
 * 
 * @param context 	Activity that instantiate a FileInfoArrayAdapter object.
 * @param values 	FileInfo[] which should be displayed/worked with.
 *
 */
public class FileInfoArrayAdapter extends ArrayAdapter<FileInfo> 
{
	private final Context context;
	private FileInfo[] values;
	
	static class ViewHolderItem{
		
		TextView tv_name;
		TextView tv_date;
		TextView tv_size;
		ImageView iv_icon;
	}
	
	/**
	 * Constructor to instantiate a FileInfoArrayAdapter.
	 * 
	 * @param context 	Activity that instantiate a FileInfoArrayAdapter object.
	 * @param resource	View that should be used by FileInfoArrayAdapter.
	 * @param values 	FileInfo[] which should be displayed.
	 */
	public FileInfoArrayAdapter(Context context, int resource, FileInfo[] values)
	{
		super(context, resource, values);
		this.context = context;				
		this.values = values;
	}

	/**
	 * method to set content of each row (one at a time).
	 * 
	 * @param position 		position (or index) of the row that should be displayed
	 * @param convertView	convertView actually never referenced but getView method needs a convertView
	 * @param parent		ViewGroup parent	
	 */
	@Override
	public View getView(int position, View convertView, ViewGroup parent) 
	{	
		ViewHolderItem viewHolder;
	
	    if(convertView==null)
	    {
	    	// inflate the layout
	    	LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
			convertView = inflater.inflate(R.layout.file_view, parent, false);
	    	
			//set up the ViewHolder
			viewHolder = new ViewHolderItem();
			
			viewHolder.tv_name = (TextView) convertView.findViewById(R.id.name);
			viewHolder.tv_date = (TextView) convertView.findViewById(R.id.date);
			viewHolder.tv_size = (TextView) convertView.findViewById(R.id.size);
			viewHolder.iv_icon = (ImageView) convertView.findViewById(R.id.icon);
			
			// store the holder with the view.
			convertView.setTag(viewHolder);
			
	    }else
	    {	    	
	    	viewHolder = (ViewHolderItem) convertView.getTag();
	    }

		// set text to the TextViews via FileInfo methods
	    viewHolder.tv_name.setText(values[position].getName());
	    viewHolder.tv_date.setText(values[position].getAcurateDate());
	    viewHolder.tv_size.setText(values[position].getAcurateSize());
		
		// check which icon should be set to the imageView
		if (values[position].isFolder()) 
		{
			viewHolder.iv_icon.setImageResource(R.drawable.folder_icon);
		} else {
			viewHolder.iv_icon.setImageResource(R.drawable.file_icon);
		}

		return convertView;
	}
} 

ListJSON/src/me/blobb/listjson/FileInfo.java
package me.blobb.listjson;

import java.util.Date;

/**
 * @author andre.boddenberg@gmx.de
 * 
 * The FileInfo class provides <i>one</i> constructor to instantiate a FileInfo object.<br>
 * A FileInfo object can be used to handle the information of folders and files.<br><br>
 * Check the parameters for more details.<br><br>
 * 
 * Besides the typical getter and setter this class provides methods, which transform the long values<br>
 * of size and date into representative String (i. e. size=>"12.5 kB"  date=>"11 Sep 1987")
 * 
 * @param size		long -> size of the FileInfo object
 * @param date		long (1.1.1970) -> date of the FileInfo object.
 * @param name		String -> name of the FileInfo object.
 * @param folder	boolean -> variable to indicate whether the FileInfo object is a folder or a file.
 */
public class FileInfo
{
	private long size;
	private long date;
	private String name;
	private boolean folder;
	
	/**
	 * Constructor to instantiate a FileInfo object.<br>
	 * There is no default Constructor, so you <i>must</i> instantiate <br>
	 * a FileInfo object with <i>all four</i> parameters.
	 * 
	 * @param size		long -> size the FileInfo object should have.
	 * @param date		long (1.1.1970) -> date the FileInfo object should have.
	 * @param name		String -> name the FileInfo object should have.
	 * @param folder	boolean -> set true if FileInfo should be a folder.
	 */
	public FileInfo(long size, long date, String name,  boolean folder)
	{
		this.name = name;
		this.size = size;
		this.date = date;
		this.folder = folder;
	}

	/**
	 * method to get a representative String of the FileInfo object.<br><br>
	 * NOTE: if you want to implement a toJSON method you could take this as a template.
	 * 
	 * @return returns a representative String of the FileInfo (may useful for testing).
	 */
	public String toString()
	{
		return "name: " + this.name + 
				", date(long): " + String.valueOf(this.date) + 
				", size (in Byte): " + this.size + 
				", is folder = " + this.isFolder();
	}
	
	/**
	 * getter: gets the size of the FileInfo object as long
	 * 
	 * @return returns the size of the FileInfo object as long
	 */
	public long getRawSize()
	{
		return size;
	}
	
	/**
	 * Method to get a representative String of the size of the FileInfo object.<br>
	 * This method divides the FileInfo size(long) by 1024 until it's smaller than 1024.<br>
	 * While it counts each division and use it as an indicator for the "unit-prefix".<br>
	 * Afterwards it rounds the size to one digit accuracy.<br><br>
	 * 
	 * NOTE: if you consider implementing "Locale" to get date in correct language, 
	 * you could use it to define whether to show a dot or a comma.
	 * 
	 * @return returns a representative String of the size of the FileInfo object (i. e. "12.5 kB")
	 */
	public String getAcurateSize()
	{
		double dSize = (double) this.size;			// casting size to double to have a floating point variable
		String dimStr = null;						// String for the "unit-prefix"
		int dim = 0;								// counts the times of dividing

		// divide size until it's smaller than 1024
		while(dSize > 1024.0D)					
		{
			dSize = dSize/1024.0D;
			dim++;
		}
		
		// evaluate the "division-counter"
		switch (dim)							
		{
			case 0:	dimStr =" B";
					break;
			case 1:	dimStr =" kB";
					break;
			case 2:	dimStr =" MB";
					break;
			case 3: dimStr =" GB";
					break;
			case 4: dimStr =" TB";
					break;
			case 5:	dimStr =" PB";
					break;
			case 6: dimStr =" EB";
		}
		
		// round to one digit accuracy
		dSize = (double)Math.round(dSize * 10) / 10;		
		
		return String.valueOf(dSize) + dimStr;
	}

	/**
	 * set the size(long) of the FileInfo object
	 * 
	 * @param size pass size of FileInfo object as long
	 */
	protected void setRawSize(long size)  
	{
		this.size = size;
	}
	
	/**
	 * getter: date of FileInfo object as long
	 * 
	 * @return returns the date as long (1.1.1970)
	 */
	public long getRawDate()
	{
		return date;
	}

	/**
	 * This method turns the date variable(long) of the FileInfo object into a representative String<br><br>
	 * 
	 * NOTE: This method can be improved a lot. Especially if you want to implement a possibility to 
	 * 		 choose what "kind/types" of the date should be displayed (and in which language).
	 * 
	 * @return returns the date of the FileInfo object as a String "DD MMM YYYY".
	 */
	public String getAcurateDate(){
		Date df = new Date(this.date * 1000);
		
		String dStr = df.toString();
		
		//String dayWord = dStr.substring(0, dStr.indexOf(" ")); 	    not used in the ListJSONTest 
		dStr = dStr.substring(dStr.indexOf(" ")+1);
		
		String monthWord = dStr.substring(0, dStr.indexOf(" "));
		dStr = dStr.substring(dStr.indexOf(" ")+1);
		
		String dayNumber = dStr.substring(0, dStr.indexOf(" "));
		dStr = dStr.substring(dStr.indexOf(" ")+1);
		
		//String time =  dStr.substring(0, dStr.indexOf(" "));			not used in the ListJSONTest 
		dStr = dStr.substring(dStr.indexOf(" ")+1);
				
		//String timezone = dStr.substring(0, dStr.indexOf(" "));		not used in the ListJSONTest 
		dStr = dStr.substring(dStr.indexOf(" ")+1);
		
		String year = dStr;
		
		return dayNumber + " " + monthWord + " " + year;
	}
	
	/**
	 * setter: sets the date of the FileInfo object
	 * 
	 * @param date pass new date as long (1.1.1970)
	 */
	protected void setRawDate(long date)
	{
		this.date = date;
	}
	
	/**
	 * getter: returns the name of the FileInfo object as a String
	 * 
	 * @return name of the FileInfo object
	 */
	public String getName()
	{
		return name;
	}
	
	/**
	 * setter: sets the name of the FileInfo object
	 * 
	 * @param name pass the name of FileInfo object (folder/file) as String
	 */
	protected void setName(String name)
	{
		this.name = name;
	}

	/**
	 * method to get information about file or folder of FileInfo object.
	 * 
	 * @return folder returns true if FileInfo object is a folder 
	 */
	public boolean isFolder()
	{
		return folder;
	}

	/**
	 * setter: set the the value of folder to "true" or "false".
	 * 
	 * @param folder
	 */
	protected void setFolder(boolean folder)
	{
		this.folder = folder;
	}

}
